kernel/src/engine/arrow_expression/evaluate_expression.rs

```diff
@@ -279,10 +279,32 @@ pub fn evaluate_expression(
             }),
             result_type,
         ) => {
-            let arrays: Vec<ArrayRef> = exprs
-                .iter()
-                .map(|expr| evaluate_expression(expr, batch, None))
-                .try_collect()?;
+            let mut arrays: Vec<ArrayRef> = Vec::with_capacity(exprs.len());
+
+            for expr in exprs {
+                let array = evaluate_expression(expr, batch, None)?;
+
+                // Short-circuit: if this array has no nulls, it is the complete result
+                if array.null_count() == 0 {
+                    // Validate result type if provided
+                    if let Some(result_type) = result_type {
+                        let expected_type = result_type.try_into_arrow()?;
+                        if array.data_type() != &expected_type {
+                            return Err(ArrowError::InvalidArgumentError(format!(
+                                "Requested result type {:?} does not match array's data type {:?}",
+                                expected_type,
+                                array.data_type()
+                            ))
+                            .into());
+                        }
+                    }
+                    return Ok(array);
```

Comment:
This isn't right, you'll just return the first array with no nulls. You just want to stop iterating at this point and not add any more arrays to `arrays`. You still need to call `coalese_arrays`. The only time you can avoid that call is if the _first_ array has no nulls.